util/mtt_update.py

```diff
@@ -0,0 +1,466 @@
+	else:
+		lOut = [f"# textdomain: {mod_name}"]
+	if header_comments is not None:
+		lOut.append(header_comments)
+
+	dGroupedBySource = {}
+
+	for key in dkeyStrings:
+		sourceList = list(dkeyStrings[key])
+		sourceString = "\n".join(sourceList)
+		listForSource = dGroupedBySource.get(sourceString, [])
+		listForSource.append(key)
+		dGroupedBySource[sourceString] = listForSource
+
+	lSourceKeys = list(dGroupedBySource.keys())
+	lSourceKeys.sort()
+	for source in lSourceKeys:
+		localizedStrings = dGroupedBySource[source]
+		if params["print-source"]:
+			if lOut[-1] != "":
+				lOut.append("")
+			lOut.append(source)
+		for localizedString in localizedStrings:
+			val = dOld.get(localizedString, {})
+			translation = val.get("translation", "")
+			comment = val.get("comment")
+			if params["break-long-lines"] and len(localizedString) > doublespace_threshold and not lOut[-1] == "":
+				lOut.append("")
+			if comment != None and comment != "" and not comment.startswith("# textdomain:"):
+				lOut.append(comment)
+			lOut.append(f"{localizedString}={translation}")
+			if params["break-long-lines"] and len(localizedString) > doublespace_threshold:
+				lOut.append("")
+
+	unusedExist = False
+	if not params["truncate-unused"]:
+		for key in dOld:
+			if key not in dkeyStrings:
+				val = dOld[key]
+				translation = val.get("translation")
+				comment = val.get("comment")
+				# only keep an unused translation if there was translated
+				# text or a comment associated with it
+				if translation != None and (translation != "" or comment):
+					if not unusedExist:
+						unusedExist = True
+						lOut.append("\n\n" + comment_unused + "\n")
+					if params["break-long-lines"] and len(key) > doublespace_threshold and not lOut[-1] == "":
+						lOut.append("")
+					if comment != None:
+						lOut.append(comment)
+					lOut.append(f"{key}={translation}")
+					if params["break-long-lines"] and len(key) > doublespace_threshold:
+						lOut.append("")
+	return "\n".join(lOut) + '\n'
+
+# Writes a template.txt file
+# dkeyStrings is the dictionary returned by generate_template
+def write_template(templ_file, dkeyStrings, mod_name):
+	# read existing template file to preserve comments
+	existing_template = import_tr_file(templ_file)
+
+	text = strings_to_text(dkeyStrings, existing_template[0], mod_name, existing_template[2], existing_template[3])
+	mkdir_p(os.path.dirname(templ_file))
+	with open(templ_file, "wt", encoding='utf-8') as template_file:
+		template_file.write(text)
+
+# Gets all translatable strings from a lua file
+def read_lua_file_strings(lua_file):
+	lOut = []
+	with open(lua_file, encoding='utf-8') as text_file:
+		text = text_file.read()
+
+		text = re.sub(pattern_concat, "", text)
+
+		strings = []
+		for s in pattern_lua_quoted.findall(text):
+			strings.append(s[1])
+		for s in pattern_lua_bracketed.findall(text):
+			strings.append(s)
+
+		for s in strings:
+			s = re.sub(r'"\.\.\s+"', "", s)
+			s = re.sub("@[^@=0-9]", "@@", s)
```

Comment:
two bugs:
1. instead of turning `@a` into `@@a` it will turn it into `@@`, losing the letter
2. `@n` works fine a translation string but isn't accepted here

Reply:
That makes me wonder. What is the need for that substitution? Doesn't the original string already need to do string escaping correctly?